### PR TITLE
export isl_ast_build_set_at_each_domain

### PR DESCRIPTION
**Pull request by skimo-openhub**

### CODE REVIEW

File: interface/cpp.cc [hunk 1]
```diff
@@ -775,15 +903,39 @@ void cpp_generator::print_destructor_impl(ostream &os,
...
+void cpp_generator::print_check_no_persistent_callback(ostream &os,
+	const isl_class &clazz, FunctionDecl *fd)
+{
+	string callback_name = clazz.persistent_callback_name(fd);
+
+	osprintf(os, "  if (%s_data)\n", callback_name.c_str());
+	if (noexceptions)
+		osprintf(os, "    return nullptr;\n");
+	else
+		print_throw(os,
```
Comment (ftynse) [inline interface/cpp.cc]: doesn't this need whitespace?

File: interface/generator.cc [hunk 2]
```diff
@@ -63,6 +65,20 @@ bool generator::is_static(const isl_class &clazz, FunctionDecl *method)
 	return extract_type(type) != clazz.name;
 }
 
+/* Does "fd" return an object that is of the same type as
+ * that of its class?
+ */
+bool generator::is_mutator(const isl_class &clazz, FunctionDecl *fd)
+{
+	ParmVarDecl *param = fd->getParamDecl(0);
```
Comment (ftynse) [inline interface/generator.cc]: what happens if the function decl has no formal arguments?

**Comment by skimo-openhub** [comment]
On Mon, Mar 19, 2018 at 02:43:25PM +0000, ftynse wrote:
> ftynse commented on this pull request.
> > @@ -775,15 +903,39 @@ void cpp_generator::print_destructor_impl(ostream &os,
>  	osprintf(os, "}\n");
>  }
>  
> +/* Print a check that the persistent callback corresponding to "fd"
> + * is not set, throwing an exception if it is set.
> + * In the C++ bindings without exceptions, simply return nullptr
> + * if the check fails.
> + */
> +void cpp_generator::print_check_no_persistent_callback(ostream &os,
> +	const isl_class &clazz, FunctionDecl *fd)
> +{
> +	string callback_name = clazz.persistent_callback_name(fd);
> +
> +	osprintf(os, "  if (%s_data)\n", callback_name.c_str());
> +	if (noexceptions)
> +		osprintf(os, "    return nullptr;\n");
> +	else
> +		print_throw(os,
> 
> doesn't this need whitespace?

You mean indentation in print_throw?
Do you think there is no enough of it?

> > @@ -63,6 +65,20 @@ bool generator::is_static(const isl_class &clazz, FunctionDecl *method)
>  	return extract_type(type) != clazz.name;
>  }
>  
> +/* Does "fd" return an object that is of the same type as
> + * that of its class?
> + */
> +bool generator::is_mutator(const isl_class &clazz, FunctionDecl *fd)
> +{
> +	ParmVarDecl *param = fd->getParamDecl(0);
> 
> what happens if the function decl has no formal arguments?

Hmm... I don't think that can happen, but it doesn't hurt to check.
I'll update the code.

Thanks.

skimo

File: interface/cpp.cc (hunk 1 above)
Comment (ftynse) [reply interface/cpp.cc]: yeah, I guess the thing that got me confused here is the printing leading whitespace here directly in `if` branch, but delegating it to `print_throw` in the `else` branch

**Comment by skimo-openhub** [comment]
On Tue, Mar 20, 2018 at 11:04:09AM +0000, ftynse wrote:
> ftynse commented on this pull request.
> 
> I have a question: since we cannot `release()` the objects that hold pointers to C++ functions and these pointers are copied when we do a copy of the object, is there a way to "detach" such pointer from the object?

Not at the moment, no.
There is also currently no way of removing these callbacks
in the C interface.  If support for that is ever requested
and implemented, the C++ interface can be adjusted accordingly.

skimo

**Comment by ftynse** [comment]
Edit: no, actually we should be able to call `.copy()` that creates a C object...

**Comment by skimo-openhub** [comment]
On Tue, Mar 20, 2018 at 11:17:09AM +0000, ftynse wrote:
> Wat worries me is that we can now create C++ objects that are no longer convertible to C objects.  If you feel it's okay, then let's go with it.

Well, ideally we wouldn't need to convert to C objects.
However, if you have a better suggestion for handling these kinds
of callbacks, then I'm all ears.

On Tue, Mar 20, 2018 at 11:18:15AM +0000, ftynse wrote:
> Edit: no, actually we should be able to call `.copy()` that creates a C object...

Yes, I do still allow those, even though it's a bit dangerous
because the user needs to ensure that the copy is only used
while the C++ object is still alive.
(This is different from .release(), which is pretty much guaranteed
to break.)

On Tue, Mar 20, 2018 at 11:17:23AM +0000, ftynse wrote:
> ftynse approved this pull request.

Thanks, I'll update the commit message wrt to your comments and
while working on some other changes, I noticed that is_mutator
should check that the method takes its first argument (and gives
its result), so I'll update that first as well.

skimo

**Pull request merged**